packages/@aws-cdk/aws-iam/lib/oidc-provider/external.ts

```diff
@@ -20,26 +21,77 @@ function defaultLogger(fmt: string, ...args: any[]) {
 /**
  * Downloads the CA thumbprint from the issuer URL
  */
-async function downloadThumbprint(issuerUrl: string) {
-  external.log(`downloading certificate authority thumbprint for ${issuerUrl}`);
+export async function downloadThumbprint(issuerUrl: string) {
+
+  external.log(`Downloading certificate authority thumbprint for ${issuerUrl}`);
+
   return new Promise<string>((ok, ko) => {
     const purl = url.parse(issuerUrl);
     const port = purl.port ? parseInt(purl.port, 10) : 443;
+
     if (!purl.host) {
       return ko(new Error(`unable to determine host from issuer url ${issuerUrl}`));
     }
+
     const socket = tls.connect(port, purl.host, { rejectUnauthorized: false, servername: purl.host });
     socket.once('error', ko);
+
     socket.once('secureConnect', () => {
-      const cert = socket.getPeerCertificate();
+      // This set to `true` would return the entire chain of certificates as a circular reference object
+      let cert = socket.getPeerCertificate(true);
+
+      const unqiueCerts = new Set<DetailedPeerCertificate>();
+      do {
+        unqiueCerts.add(cert);
+        cert = cert.issuerCertificate;
+      } while ( cert && typeof cert === 'object' && !unqiueCerts.has(cert));
+
+      // The last `cert` obtained must be the root certificate in the certificate chain
+      const rootCert = [...unqiueCerts].pop()!;
+
+      // Add `ca: true` when node merges the feature. Awaiting resolution: https://github.com/nodejs/node/issues/44905
+      if (!(util.isDeepStrictEqual(rootCert.issuer, rootCert.subject))) {
+        return ko(new Error(`Subject and Issuer of certificate received are different. 
+        Received: \'Subject\' is ${JSON.stringify(rootCert.subject, null, 4)} and \'Issuer\':${JSON.stringify(rootCert.issuer, null, 4)}`));
+      }
+
+      const validTo = new Date(rootCert.valid_to);
+      const certificateValidity = getCertificateValidity(validTo);
+
+      if (certificateValidity < 0) {
+        return ko(new Error(`The certificate has already expired on: ${validTo.toUTCString()}`));
+      }
+
+      // Warning user if certificate validity is expiring within 6 months
+      if (certificateValidity < 180) {
+        /* eslint-disable-next-line no-console */
+        console.warn(`The root certificate obtained would expire in ${certificateValidity} days!`);
```

Comment:
Should this warning use `Annotations.of(this).addWarning()` instead of the console statement? You can also write a test for that to ensure the warning is being emitted.

Reply:
Looks like this is associated with a construct. Since this is a function called within a lambda handler, `this` value is not present here. Let me know if I am understanding this incorrectly. 

`Annotations.of(this).addWarning(`The root certificate obtained would expire in ${certificateValidity} days!`);`

Reply:
What context is this function used in? Is it just something users can use directly in their app? If so, you're right that the annotation can't be used. I had assumed it was used in the context of a construct or stack in our library. If that's the case, you can use a bind function, but now I'm thinking that may not be the case so you can discard this comment as well.

Reply:
So this function is being used by a custom resource [here](https://github.com/aws/aws-cdk/blob/main/packages/@aws-cdk/aws-iam/lib/oidc-provider.ts#L152). Let me know if you know this can work in this scenario.